modules/core/src/system.cpp

```diff
@@ -1113,10 +1124,23 @@ class TlsStorage
         return NULL;
     }
 
+    // Gather data from threads by TLS storage index
+    void gather(size_t slotIdx, std::vector<void*> &dataVec) const
+    {
+        CV_Assert(tlsSlots.size() > slotIdx);
+
+        for(size_t i = 0; i < threads.size(); i++)
+        {
+            std::vector<void*>& thread_slots = threads[i]->slots;
+            if (thread_slots.size() > slotIdx && thread_slots[slotIdx])
+                dataVec.push_back(thread_slots[slotIdx]);
+        }
+    }
+
     // Set data to storage index
     void setData(size_t slotIdx, void* pData)
     {
-        CV_Assert(pData != NULL);
+        CV_Assert(tlsSlots.size() > slotIdx && pData != NULL);
```

Comment:
Below:
```
if(slotIdx >= threadData->slots.size())
    threadData->slots.resize(slotIdx+1);
```
I beleive this code should be also guarded (via scheeme: check / lock / check again / do job)